### ISSUE TITLE
ENH: randomized svd and eigh decomposition (faster but approximate)

### ISSUE THREAD

**Issue by smarie** [scipy/scipy#23145]
### Is your feature request related to a problem? Please describe.

_(This feature refers to scikit-learn as the "first user" but applies to other scientific models and algorithms)_

**Context**: Many _classic Machine Learning models_ rely on singular value decomposition (SVD) or eigenvalue decomposition (EIG), often on real symmetric or hermitian matrices (EIGH, H meaning "hermitian"). In many situations only the first k singular values or eigenvalues and associated singular vectors or eigenvectors are required.

Scipy currently supports those algorithms through the following routines:

- EIGH : with [LAPACK](https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.eigh.html#scipy.linalg.eigh), with [ARPACK](https://docs.scipy.org/doc/scipy/tutorial/arpack.html), with [Lobpcg](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.lobpcg.html#scipy.sparse.linalg.lobpcg)
- SVD : with [LAPACK](https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.svd.html#scipy.linalg.svd), with [ARPACK and LOBPCG](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.svds.html#scipy.sparse.linalg.svds)

They (not always all of them) are available as solvers in [FastICA](https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.FastICA.html), [KernelPCA](https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.KernelPCA.html), [PCA](https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.PCA.html), [LDA](https://scikit-learn.org/stable/modules/generated/sklearn.discriminant_analysis.LinearDiscriminantAnalysis.html), [Isomap](https://scikit-learn.org/stable/modules/generated/sklearn.manifold.Isomap.html), [NeighborhoodComponentsAnalysis](https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.NeighborhoodComponentsAnalysis.html), [RidgeCV](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.RidgeCV.html), [RidgeClassifierCV](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.RidgeClassifierCV.html), [LocallyLinearEmbedding](https://scikit-learn.org/stable/modules/generated/sklearn.manifold.LocallyLinearEmbedding.html), [SpectralEmbedding](https://scikit-learn.org/stable/modules/generated/sklearn.manifold.SpectralEmbedding.html), [SpectralClustering](https://scikit-learn.org/stable/modules/generated/sklearn.cluster.SpectralClustering.html).

To complete the picture, 

- A PR is open to add LAPACK `eigh` solver to [MDS](https://github.com/scikit-learn/scikit-learn/pull/22330)
- a solver named [`amg`](https://pyamg.readthedocs.io/en/latest/) (not in scipy), is used in [SpectralEmbedding](https://scikit-learn.org/stable/modules/generated/sklearn.manifold.SpectralEmbedding.html) and [SpectralClustering](https://scikit-learn.org/stable/modules/generated/sklearn.cluster.SpectralClustering.html). I found [this past open discussion about it](https://github.com/scikit-learn/scikit-learn/issues/14713).

**Problem**: When matrices become large, usually because the number of samples and/or their dimensionality become large, execution time increases drastically. Lobpcg and Arpack sometimes do not lead to fast enough results.

**Current solutions**

1. **Parallel computing** (out of scope of this feature request) : Exact decomposition is standardized in the Array API : [EIGH](https://data-apis.org/array-api/2023.12/extensions/generated/array_api.linalg.eigh.html#array_api.linalg.eigh), [SVD](https://data-apis.org/array-api/2023.12/extensions/generated/array_api.linalg.svd.html), and is therefore parallelizable on GPU depending on implementation (numpy, cupy, torch, dask, jax, pydata sparse...). Scikit-learn has been adding such support in the recent releases, for an increasing number of ML estimators.

2. **Approximate solvers** (this feature request) : In scikit-learn we have introduced alternative solvers for some models, to get faster results with good approximations. The family of such solvers is provided in [“Finding structure with randomness: Stochastic algorithms for constructing approximate matrix decompositions”](https://arxiv.org/abs/0909.4061) Halko, et al., 2009. 

    - Algorithm 4.3 from this paper has been successfully introduced for [PCA](https://scikit-learn.org/stable/modules/decomposition.html#pca-using-randomized-svd) and for [Kernel PCA](https://scikit-learn.org/stable/modules/decomposition.html#choice-of-solver-for-kernel-pca). [Execution results are quite good](https://github.com/scikit-learn/scikit-learn/pull/12069#issuecomment-430270699)
    - Algorithm 5.3 from the same paper has recently been proposed for [KernelPCA and Isomap](https://github.com/scikit-learn/scikit-learn/pull/31247), with promising [execution results (right-side of this figure)](https://github.com/scikit-learn/scikit-learn/pull/31247#issuecomment-2832290795). Once this is merged, we will propose similar additions for some of the estimators listed above (e.g. RidgeCV, MDS..)

### Describe the solution you'd like.

It could be worth 

- moving [`randomized_svd`](https://scikit-learn.org/stable/modules/generated/sklearn.utils.extmath.randomized_svd.html) and [`randomized_eigh`](https://github.com/scikit-learn/scikit-learn/pull/31247/files#diff-86c94a3ca33490c6190f488f5d40b01bf0fd29be36da0b4497ef0da1fda4148aR692) implementation to SciPy, and making sure they are cross-referenced from their "exact" (svd, eigh) and "sparse" (svds, eigsh) equivalents' documentation. This would bring immediate benefits 

    - to the broader scientific community using scipy, as practitioners will see that such solvers exist and will be in a position to try them without any effort (same scipy package, same API signature)
    - to the scikit-learn maintainers, as they will more easily be in a position to discuss issues and maintenance of such solvers with the broader scipy community

- improving the SciPy documentation so that it is clear for non-experts that these solvers (and maybe many others) are approximate. Could we possibly imagine a distinct `scipy.approx.linalg` package ?

    - `scipy.linalg`: existing exact decomposition
    - `scipy.sparse.linalg`: existing for sparse matrices
    - `scipy.approx.linalg`: faster but with lower precision

In the future we could imagine that the visibility of such a package would encourage researchers in the field to propose additional methods.

### Describe alternatives you've considered.

Not an alternative but rather an existing link with a past discussion in SciPy: looking around I figured out that such approaches were already mentioned as interesting in 2017 by a SciPy user in the past https://github.com/scipy/scipy/issues/7561, and in the dicussion he mentioned an implementation in [pyscgm](https://github.com/dsuess/pyscgm) and @rgommers mentioned [`scipy.linalg.interpolative`](https://docs.scipy.org/doc/scipy/reference/linalg.interpolative.html#module-scipy.linalg.interpolative) and [_sketches](https://github.com/scipy/scipy/blob/main/scipy/linalg/_sketches.py) (which seems to correspond to [this linalg documentation section](https://scipy.github.io/devdocs/reference/linalg.html#sketches-and-random-projections)).

### Additional context (e.g. screenshots, GIFs)

Thanks for reading this very long feature description ! And please pardon me for any mistake in the above details.

**Comment by ilayn** [comment]
See this discussion too https://discuss.scientific-python.org/t/add-lsrn-solver/1888/1

**Comment by smarie** [comment]
Very relevant indeed @ilayn ! thanks for the link

**Comment by lobpcg** [comment]
The AMG solver mentioned in the description actually calls scipy LOBPCG to my recollection but has an extra special feature of building a  preconditioner for LOBPCG not included in scipy. 

The construction of the preconditioner in AMG is based on sketching, so LOBPCG with AMG preconditioning is a direct competitor for purely random sketching algorithms .

**Comment by lobpcg** [comment]
Both ARPACK and LOBPCG are in fact randomized algorithms since they by default use random initial approximations. 

Moreover, LOBPCG directly competes with the randomized codes from sk-learn being simply run a small number of iterations making it extremely fast but producing only vague approximations. In other words, the randomized codes from sk-learn might be just inferior compared to LOBPCG in all scenarios and thus not worth the time moving to scipy.

LOBPCG is basically a sketching on Krylov steroids.

**Comment by ilayn** [comment]
Randomized algorithms here mostly refer to sketching and the related algorithms. ARPACK does not really do any sketching but tries to discover a Krylov basis. I think that's a bit of overgeneralization but taxonomy wise I agree. If I read the intention here correctly they are looking for something like https://github.com/BallisticLA/RandBLAS

**Comment by lobpcg** [comment]
See also 7yo discussion https://github.com/scikit-learn/scikit-learn/issues/12068

"Everything new us just well forgotten old"  😃

**Comment by ilayn** [comment]
Not sure I follow what you mean. 

What I meant is RNLA are mostly meant for massive operators where you can't even form a basis or perform any Arnoldi/Lanczos-like iterations. So you work with sketches exclusively. Hence ARPACK/LOBPCG/AMG are not suitable for such workloads. This issue also mentions the same limitation.

That's what the issue description mentions about the approximate solvers which is similar to what RandBLAS tries to solve.

Heiko's paper and the recent methods are not old in my book.

**Comment by lobpcg** [comment]
> Not sure I follow what you mean. 
> 
> What I meant is RNLA are mostly meant for massive operators where you can't even form a basis or perform any Arnoldi/Lanczos-like iterations. So you work with sketches exclusively. Hence ARPACK/LOBPCG/AMG are not suitable for such workloads. This issue also mentions the same limitation.
> 
> That's what the issue description mentions about the approximate solvers which is similar to what RandBLAS tries to solve.
> 
> Heiko's paper and the recent methods are not old in my book.

Sketching means multiplication of the matrix by a block of random vectors  - that is exactly the first iteration of LOBPCG. The difference is that sketching just stops there while LOBPCG iterates if better accuracy is required. 

LOBPCG  doesn't form and store any Lanczos-like basis but instead works with sketches exclusively although they are no longer just random giving  LOBPCG better accuracy per costs compared to the purely random sketching.

**Comment by ilayn** [comment]
Sketching is about representing the operator with smaller size with probabilistic guarantees. The point is RNLA is not just about eigenvalues but many other BLAS/LAPACK-like operations again with probabilistic guarantees on the accuracy. We can and do use these in other methods too but at least for arpack it is a hit or miss with not much a priori guarantees. 

If you are interested, try arpack or lobpcg on arrays with sizes above 32bit limit to see the performance or feasibility of the methods though you need 64bit BLAS/LAPACK to start with.
